### ISSUE TITLE
Switch to bootstrap-table for tabular views (replacing manage_tables.js, search, filters, action buttons)

### ISSUE THREAD

**Issue by daN4cat** [jekkos/opensourcepos#293]
Refactor manage_tables.js such we can move more code contained in other page in the common js script and easy maintenance/reusability (as mentioned in #120).

Alternatively we could consider changing the tabular view with something that matches better CI3 and bootstrap, something like in #157 by jekkos:

>Looks like CI3 has some nice REST extensions. Regarding a possible redo of the application, tables might be easier to do using REST + react.js bootstrap table component (providing pagination, bootstrap layout, all sorts of filtering options, by just configuring a couple of things)

>http://torbjornzetterlund.com/how-to-work-with-restful-services-in-codeigniter-3-x/
>https://github.com/AllenFang/react-bootstrap-table

>Also some services should then be restified for the android app :)

**Comment by jekkos** [comment]
Ok this is a fun one. Will try to read up a bit on react.js when I find the time.

**Comment by daN4cat** [comment]
Maybe we can split in two, one part with just a bit of refactoring to move to manage_tables.js the common code and leave the replacement of tabular view for a second step.

**Comment by jekkos** [comment]
Another option for the tables might be folowing pointer. Demo's look quite nice.

https://github.com/wenzhixin/bootstrap-table

**Comment by daN4cat** [comment]
Yes, we need to look for bootstrap js solutions.

I used the demo from the link about and it's buggy. Type in the search box Item and then if you add space and 9 it rewrites to Item9 and doesn't find any item....

**Comment by daN4cat** [comment]
Beside that it looks nice with different views and export modes. Really cool.

Does it support ajax for remote query?

**Comment by daN4cat** [comment]
This is probably part of #328 too.

**Comment by daN4cat** [comment]
I think we should replace this with a switch to https://github.com/wenzhixin/bootstrap-table

**Comment by jekkos** [comment]
Ok so I'll be working on this story and you on the other? What about 2.4 shouldn't we try to finish issues and get it ready to ship? So we can fully focus on this branch?

**Comment by daN4cat** [comment]
Yes, let's finish off 2.4 and the ones that are currently open.

About this, rework or register and change flash to jchart or jsPlot, maybe those can be a 3.1 story?

**Comment by jekkos** [comment]
hmm maybe I'd like to have the tables in 3.0 so we have a full rework with bootstrap? Maybe the charts aren't that much work either.. Currently it's the ofc library that's used for that.

**Comment by daN4cat** [comment]
Yes, we can change this once 2.4 and 3.0 are sorted. We can move back to 3.0 what's 3.1.
It's just to prioritise things at the moment.

**Comment by ais-one** [comment]
Hi,

You may not want to go down the reactJS path, a lot of headaches.

You may want to consider using RactiveJS instead. I have been doing research on using these stuff for CRUD, eventually, I realized that it is really a different world from using CI.

You may want to just use Bootstrap CSS to make the layout responsive.

Or use something like lazy_mofo for CRUD.

**Comment by jekkos** [comment]
OK thanks for sharing these thoughts. Did you work with react before? Also which kind of problem did you encounter?

I just react js as a way to create components from recurrent building blocks and make them reusable across the application. Also it's view technology only so it shouldn't impact our CI code. Only thing that changes is that all data will need to be loaded as json (thus no php echo of html or the like).

**Comment by ais-one** [comment]
Well, reactJS has this thing called JSX. Anyway their performance (ReactJS, EmberJS, AngularJS is slow).

You want fast and easily maintainable, look at MithrilJS or RactiveJS.

Also if going down that route, javascript will be used to setup the custom inputs of the columns (such as selects, etc.), rather than PHP spitting out those customer columns.

Actually, with good pagination, PHP a solution is still ok for CRUD, despite spitting out overhead...

**Comment by daN4cat** [comment]
BTW, bootstrap-table can be used with a json interface, right?
Not sure we need to revolutionise that much things here.

**Comment by ais-one** [comment]
yes the bootstrap table css can be used with JSON interface, those JS thingys will form the table for you.

for simple tables it is easy, but more complex still need to code abit more in JS.

some of those JS stuff are pure over-kill and over-engineering.

**Comment by jekkos** [comment]
Ok fact that we can get rid of table_helper is already good. I don't really like the way views are rendered in the first place. After all we're doing ajax with plain json to update individual rows already. Same for search. Why not do that for initial table buildup as well? Sounds more consistent in the end.

Maybe you're right on the over engineering, however these frameworks should provide a better way to abstract and organise code. On the other hand it will probably be harder to tackle specific issues if the used abstractions are leaky. Think bootstrap tables will do for now indeed as the jsx would need additional support from a build system (something like babel with gulp or grunt) to render the actual js.

**Comment by ais-one** [comment]
I am for anything to better organise code.

Which is why I had mentioned RactiveJS or MithrilJS if you want to push the work to the client side.

Performance and maintainability wise, I would go for either of them. MithrilJS is ridiculously light though.

**Comment by daN4cat** [comment]
I think the first step is to move all the class/functions from controllers connected to the webbrowser UI to use json. Controllers are still not like that in different parts.
Once the json I/O is in place using standard ajax with Bootstrap plugins will do. Please note that we are talking of this bootstrap-table: https://github.com/wenzhixin/bootstrap-table, not what's provided standard.

Now I don't know if any of the additional light frameworks will do any better.
But I'm wary about adding many paradigms here if we can keep it simple. The chance that somebody can help going forward reduces if you "got too fancy".
I'm speaking by experience, I had in the past situations where people go so "guru" that you are left with a mess when they leave.
We can see already here how some people come asking basic things and they get lost with CI paradigms. Figure out the rest...

**Comment by ais-one** [comment]
Ah I see, I thought you were considering to use the standard bootstrap table with something like reactJS.

bootstrap-table seems pretty well supported, I think it is sufficient for the purpose.

**Comment by jekkos** [comment]
I've started to play around with bootstrap tables this morning. It's actually really easy to get started with an existing table as well! just a matter of removing the sortable class and adding a data-toggle attribute to the table tag! This simple change will restyle the whole table and give it a nice modern look.

Also sorting is really easy to add, just a matter of adding some data- attributes to the table headers.

Would it be ok if I try to add the new button to the button group shown right above the table in most of the examples? Same for the import button. WIll give a cleaner and more compact look after all.

**Comment by jekkos** [comment]
As most of these changes don't require too much effort at first sight, I'd like to have them in 3.0 as well (inlcuding removal of obsolete code in manage_table.js)

**Comment by daN4cat** [comment]
Buttons with group action, filters and search should all go on the top of the table header. I had in mind to to move the filters to a multiselect drop down to save space.

**Comment by daN4cat** [comment]
Detail sale report is different because supporting table nesting.

**Comment by jekkos** [comment]
Also the excel import shouldn't we move it to the button group? Won't if affect users with touchscreen?

Same for the new button. Maybe we can add it to the table header as the filters will move anyway?

**Comment by daN4cat** [comment]
What I meant is two rows:

- 1st row: pagination on the left, new/import/etc blue buttons on the right
- 2nd row: group action (grey) buttons on the left, filters after, search on the right

**Comment by daN4cat** [comment]
I'm trying to work on the multiselect/dropdown for filters at the moment. I have it on the view but I cannot read the values in the post. Any clue how to read the filter list of a multiselect on a post in the controller?

**Comment by daN4cat** [comment]
I've probably found out now...

**Comment by daN4cat** [comment]
OK, the multiselect is sorted. There are now all the rules around the search according to various conditions.

**Comment by jekkos** [comment]
Ok in that case I'll leave that alone and focus on the table + sorting

**Comment by daN4cat** [comment]
Here we go, changed the filter part in Items as above mentioned. See if you like it... we are a step closer to get the end.

**Comment by daN4cat** [comment]
And this change was not a trivial one as it took a day to get everything to work fine.

**Comment by odiea** [comment]
So the way it now works is you use the dropdown box and select the filter . then you have to click back on the dropdown box to get the filter to work?

**Comment by daN4cat** [comment]
Yes, you do multiple selections and when finished you click outside and the filter kicks in.

**Comment by odiea** [comment]
OK Thanks. Fancy now isn't it.

**Comment by odiea** [comment]
And how do we now add new filters if needed to show in the dropdown?

**Comment by daN4cat** [comment]
In Items controller:

- in function index you add in data['filters'] array the item '<name>'=>$this->lang->line('<name translation>')
- in function search you just need to add the '<name>'=>FALSE

Then of course in the model Item function search you need to handle the $filters addition

**Comment by odiea** [comment]
Thanks. I will try to add my printed feature in there.

**Comment by jekkos** [comment]
![screenshot from 2016-03-30 22 54 08](https://cloud.githubusercontent.com/assets/576234/14157261/8a4ce9aa-f6ca-11e5-83f8-7f1f6d339f10.png)

**Comment by jekkos** [comment]
this is how customers currently looks like in my build. Still need to fix up the refresh after edit part. Also animation on delete is broken now.

**Comment by daN4cat** [comment]
@jekkos looks promising, I bet that you have a lot to test to make sure everything works fine in all possible combinations.

**Comment by daN4cat** [comment]
I my last commit, I applied the same changes to Sales/Takings and so we said goodbye to the old filters block forever. I tidied up the stylesheet as well.

I didn't change the daterange to be the same as items because Takings requires time and in the end it's working fine. Maybe one day I'll change for the daterangepicker, it's fine now.

**Comment by odiea** [comment]
Looks and works great.

**Comment by jekkos** [comment]
Maybe create an issue for the daterangepicker, I'll look into it after I finish the tables.

**Comment by daN4cat** [comment]
Is it really an issue? It's working fine with what it is at the moment and fully bootstrap-look.
Not sure we want to spend more time on this now? It's more a nice to have.

The issue with the change is that we would need to add the time support in the datarangepicker and get all the conversions done right.

**Comment by jekkos** [comment]
Ok but do we really need the timestamps in the takings overview? In my case I think it's more of a nuisance. I think it's sufficient to filter by date and thus speed up the filter selection a bit.

Also I think it's more user friendly to force a user to select the date range all at once, without having to open yet another dialog to enter the end date

**Comment by daN4cat** [comment]
Pull the filters to the right side before the search to make a distinction between what is action buttons (left) and filters +seach (right). Also style differs to make things neat.

**Comment by daN4cat** [comment]
@jekkos re your question, as said this works fine and most of the time is "today". The time is to fine tune searches in case people want to narrow down on range related to a payment transaction.

**Comment by jekkos** [comment]
Been playing some more with bootstrap tables. Fact is that they allow to add some extra filters real easy just by adding some options to the table itself.

But this interferes with our own filters.. So will need to find a way to integrate them both in a nice way.

**Comment by jekkos** [comment]
Also found the documentation on the bootrstrap tables a bit cumbersome. Some behavior isn't well described which forces me to look into the code to find how thing should work.. taking quite a bit more time than expected

**Comment by daN4cat** [comment]
Shall we move it back to 3.1? What we have at the moment would do in 3.0. I would like to keep new table separate.

**Comment by daN4cat** [comment]
I was expecting it not to be an easy walk.. .

**Comment by jekkos** [comment]
It's ok just takes a bit more time to figure out how to replace the existing code with what is included in the tables. It's just a typical sort of problem with these libraries, I had it with jquery ui as well. It's just faster to read their source than to go through their docs..

**Comment by jekkos** [comment]
Once i have one module working it will be a matter of find replace for (most) of the others so guess that part will go faster

**Comment by daN4cat** [comment]
Just checking bootstrap-table and would be nice to have the option to select which column to display and maybe storing the layout information.
In this way people could add other fields to the table without big issues.

I was even wondering about the filter / search by column that could be an option instead of the filter approach we have now? Just some thoughts....

**Comment by daN4cat** [comment]
Even the in-cell editing for cases like items looks a nice feature.

**Comment by jekkos** [comment]
Yes I already have this functionality in place, configurable columns. Also the table has search builtin. It might be able to superseeed our search, but then we'll need to load the whole resultset in memory. Might hurt performance once item or customers list is getting bigger

**Comment by daN4cat** [comment]
Preload everything is a no go I'm afraid. Selected deleted would load probably 20000 items, I guess oidea case would be even worst.

**Comment by jekkos** [comment]
I have restarted this implementation, got rid of the get_all in model and just use search in controllers with ajax. Also tables are initialized now using javascript only and always loaded asynchronously even the first time the page loads. A lot of old functionality can be directly replaced with what is built in in this plugin (row selection, show search field, pagination refresh cycle, ..)

I enabled server side pagniation so this part so this can be reused from older code. One problem is that client side sorting is disabled in that case. I need to look further whether column search cannot make an ajax call that just refreshes the table using a new order by...

**Comment by daN4cat** [comment]
Have you changed table_helper too? Or got rid of it?

**Comment by jekkos** [comment]
Think we should keep it, as some methods are reused across controllers / reports. This monring i realized that reports tabular views should be changed as well. Bootstrap tables can handle embedded tables properly so that shouldn't really be an issue. 

Possible bug in current: do dialogs still work after changing pagination offset? I had to change the dialog support init because data is loaded through ajax now. Think this probably needs to be done in current version as well after clicking a page number in the pagination links (haven't tested yet)

**Comment by jekkos** [comment]
Stilll working on this one, today I fixed some bugs in customer section. I'm also reviewing most of the existing code in manage_table and moving the new stuff to a table_support object (similar to the dialog_support we have for dialogs). This way the code is more 'object oriented' and reusable in the views.

**Comment by jekkos** [comment]
Just pushed my working branch to this repo (for backup purposes). I'm considering to drop the search suggestions in the tables as we can enable immediate column search anyway.

**Comment by daN4cat** [comment]
you mean search directly on each column?

**Comment by odiea** [comment]
@jekkos  I just tried your version and I am getting this. There is also no search results in customer Register.
Fatal error: Call to undefined function get_people_manage_table() in C:\xampp\htdocs\bootstraptables\application\controllers\Employees.php on line 17
A PHP Error was encountered
Severity: Error
Message: Call to undefined function get_people_manage_table()
Filename: controllers/Employees.php
Line Number: 17
Backtrace:

**Comment by jekkos** [comment]
@odiea Thanks for testing, it's still a wip, need to review and test some more stuff indeed, not quite finished yet :)
@daN4cat yes now the search is called on keypress immediately. I guess in this case there's no more need to show suggestions as the table updates in real time (using ajax)

**Comment by daN4cat** [comment]
sounds cool

**Comment by jekkos** [comment]
Ok another intermediate commit here. There are probably some things not working at this moment but I'm focusing on getting the customers module done first as an example for the rest. Most of the heavy js lifting is done which will speed up the refactor for the other modules.

* table update/delete/create tested and working for customers with bootstrap tables
* general cleanup of the html (remove ugly old constructs)
* using proper js namespacing now for ospos js code
* 'real time' search (without suggestions) should work using HTTP GET with only search function in the backend (got rid of the backend code for initial load)
* add some glyphicons for buttons in customer module
* use disabled state for delete/email if these actions are not applicable

**Comment by jekkos** [comment]
Haven't tested pagination yet but I guess we will need to choose between builtin bootstrap-tables (that offers pagination by adding query string parameters) or use our own.

**Comment by daN4cat** [comment]
Question: would the new way of avoiding the suggestion work with partial word search?

**Comment by daN4cat** [comment]
With regards to pagination, not sure but maybe using what comes with bootstrap-tables is better?
In the end we use it only with tabular views and using all from bootstrap-tables will make things more consistent.
Provided it doesn't require fetching all the data but still works with ajax / controller/search.

**Comment by jekkos** [comment]
@daN4cat partial word search should be fine. Bootstrap tables triggers a search after each letter change. It does feel a little slower maybe and a waiting animation or something on the table could be nice to indicate that a search is happening. Currently I've kept the autocomplete code and it shouldn't be too much work to revert to the old situation for any reason.

**Comment by jekkos** [comment]
I found myself able to cleanup quite a bit of code (and I'm not done yet) so yes from that perspective I guess it would be better to use bootstrap table's pagination. Most of the supplier/customers view look quite simple now. Also there's some more cleanup to be done in the controllers as code could be reused there as well.

**Comment by daN4cat** [comment]
Yes, the more we reuse the less to maintain or add if we create something new. As said before I found this rework as something where I remove more than add.
Good to keep things lean.
It was a good idea to kick start 3.

**Comment by jekkos** [comment]
I've started the work on items module, a few difficulties there

* There's no standard way to add the controls for this table to both left and right side. MIght be able to make it look as it is now with a bit of js hacking (or bootstrap-tables modifications) but will try to make the whole work first
* There's no supported way to add the extra search parameters on top of the standard order, limit and search ones. One can only choose between sending his own or using the ones builtin. WIll probably need to come up with some kind of hack to make it work

**Comment by daN4cat** [comment]
Maybe you can contribute back bootstrap-tables any addition so it's easier to maintain going forward.

The filters is like an advance search, so surprised there is no support.

**Comment by jekkos** [comment]
Found a workaround for the first point. I have the items list loading now. Still need to work a bit on the filter handling. Toolbar layout is screwed up currently but I'll try to find a workaround to have everything in one line. 

Wouldn't it also be nice to have the item filters, stock locations and date range picker in one[ button group](http://designshack.net/wp-content/uploads/bootstrap2-9.jpg)?

**Comment by daN4cat** [comment]
Try to have all in one line and see if it works.

**Comment by jekkos** [comment]
Yes sure first make it work, improve later on.

**Comment by jekkos** [comment]
Think I found a way to customize the toolbar. There's a couple of [ 'extensions' ](https://github.com/wenzhixin/bootstrap-table/blob/develop/dist/extensions/filter-control/bootstrap-table-filter-control.js) to show how to do that properly.

**Comment by jekkos** [comment]
Ok continues the work a little bit. Will need further fix item form handling tomorrow, especially the new item forms in receivings and sale register.

Also remaining to be done 
* enable localisation
* move table init to table_helper
* remove init_delete function and get translation from csv files
* migrate takings overview
* migrate tables in reports

**Comment by jekkos** [comment]
Fixed some of the items form handling now. inventory, count_details should all update the table properly. Was also reworking the bulk edit handling (think it's done but haven't tested). Barcode generation works AFAIK but haven't tested pagination.

Maybe I will need to refactor the js to use prototypal inheritance for table_support so that functions can be easily extended/overridden in the respective modules.

**Comment by daN4cat** [comment]
Just a question, are you rebasing your development with the latest master?
I'm worried that you might change parts that had development and then the merge will be painful or prone to errors.

**Comment by jekkos** [comment]
I haven't done any rebase since a couple of days. I didn't want to be disturbed with any upstream changes and will rebase once my changes are done and tested. 

Think it'll be best to integrate master in the end and ask for some testing. If there are no more bugs then we can move to master.

**Comment by daN4cat** [comment]
The more you wait the more merge time. I guess you know already this by experience.

Anyway, yes first rebase/merge in the branch for testing and review and then of course release to master.
Master is pretty stable and usable as it is, so we should keep it like that as I'm adding features bit a bit.

**Comment by jekkos** [comment]
Ok did some more fixing on the items module (bulk edit) + fixed problems in item kits (should work ok now).

Does anyone know whether there was a delete functionality in item kits? I noticed that there is no deleted column in that table so it might have never worked :)

**Comment by jekkos** [comment]
So I have two options:

* Omit the delete button in item kits
* Add the deleted column
* Reuse deleted column from items

**Comment by jekkos** [comment]
Ok delete button seems to be present. Funny that this bug was never reported. I

**Comment by daN4cat** [comment]
Well, when I changed stuff I found that we had bugs there from old versions but nobody ever bother or managed to report.

This structural refactoring if improving things under different aspects.

**Comment by odiea** [comment]
in items table there is no longer filtering for stock locations.

**Comment by jekkos** [comment]
Ok progressed this part some more 

* All of the modules working now (there might still be bugs)
* Fixed new item, customer and supplier in register / receiving
* Add glyphicons to most buttons
" Further refactor / cleanup of takings

Left todo

*   enable localisation
*   migrate tables in reports

**Comment by jekkos** [comment]
Think it's about time to sync branches..

**Comment by daN4cat** [comment]
Yes, you must rebase because I made sevaral changes and there will be merge conflicts.

**Comment by jekkos** [comment]
Just rebased with master until the 10th of april.. still some commits to go but will continue with those later (quite some new stuff in the master since I started on this feature)

**Comment by daN4cat** [comment]
Yes, I needed to progress in master. As suggestion, once you have completed with the rebase do a merge at the end on your branch to make sure you have not fail a rebase conflict resolution.

**Comment by jekkos** [comment]
I synced the branch now with latest master and have deployed a sanity checked version oin my test server. 

Would be nice If anyone could take some time to play around to check for issues / regressions. Next thing for me is fix the detailed reports (which are probably broken now)

**Comment by daN4cat** [comment]
I will try it this eve

**Comment by daN4cat** [comment]
And thank you for all this work!

**Comment by jekkos** [comment]
Wil lclose this thread for now (it became too long). Testing phase has begun so I will probably discuss issues in #507 instead.

**Comment by jekkos** [comment]
Bootstrap tables localisation is now integrated with ospos language file system (also #439)

**Comment by jekkos** [comment]
Think this one is finished now?
